src/System.Private.ServiceModel/src/System/IdentityModel/CryptoHelper.cs

```diff
@@ -61,7 +97,305 @@ internal static bool IsAsymmetricAlgorithm(string algorithm)
+                    found = true;
+                // The reason we do not return here even when the user has provided a custom algorithm to CryptoConfig 
+                // is because we need to check if the user has overwritten an existing standard URI.
+            }
+
+            switch (algorithm)
+            {
+                case SecurityAlgorithms.DsaSha1Signature:
+                case SecurityAlgorithms.RsaSha1Signature:
+                case SecurityAlgorithms.RsaSha256Signature:
+                case SecurityAlgorithms.RsaOaepKeyWrap:
+                case SecurityAlgorithms.RsaV15KeyWrap:
+                    return false;
+                case SecurityAlgorithms.HmacSha1Signature:
+                case SecurityAlgorithms.HmacSha256Signature:
+                case SecurityAlgorithms.Psha1KeyDerivation:
+                case SecurityAlgorithms.Psha1KeyDerivationDec2005:
+                    return true;
+                case SecurityAlgorithms.Aes128Encryption:
+                case SecurityAlgorithms.Aes128KeyWrap:
+                    return keySize >= 128 && keySize <= 256;
+                case SecurityAlgorithms.Aes192Encryption:
+                case SecurityAlgorithms.Aes192KeyWrap:
+                    return keySize >= 192 && keySize <= 256;
+                case SecurityAlgorithms.Aes256Encryption:
+                case SecurityAlgorithms.Aes256KeyWrap:
+                    return keySize == 256;
+                case SecurityAlgorithms.TripleDesEncryption:
+                case SecurityAlgorithms.TripleDesKeyWrap:
+                    return keySize == 128 || keySize == 192;
+                default:
+                    if (found)
+                        return true;
+                    return false;
+                    // We do not expect the user to map the uri of an existing standrad algorithm with say key size 128 bit 
+                    // to a custom algorithm with keySize 192 bits. If he does that, we anyways make sure that we return false.
+            }
+        }
+
+        internal static void FillRandomBytes(byte[] buffer)
+        {
+            RandomNumberGenerator.GetBytes(buffer);
+        }
+
+        /// <summary>
+        /// This generates the entropy using random number. This is usually used on the sending 
+        /// side to generate the requestor's entropy.
+        /// </summary>
+        /// <param name="data">The array to fill with cryptographically strong random nonzero bytes.</param>
+        public static void GenerateRandomBytes(byte[] data)
+        {
+            RandomNumberGenerator.GetNonZeroBytes(data);
+        }
+
+        /// <summary>
+        /// This method generates a random byte array used as entropy with the given size. 
+        /// </summary>
+        /// <param name="sizeInBits"></param>
+        /// <returns></returns>
+        public static byte[] GenerateRandomBytes(int sizeInBits)
+        {
+            int sizeInBytes = sizeInBits / 8;
```

Comment:
Are your bits guaranteed multiples of 8?  I recommend either asserting they are, or using `(sizeInBits + 7) / 8)`, so 0 => 0, 1 => 1, 7 => 1, 8 => 1, 9 => 2, etc.

Reply:
I believe the else block covers this as it multiplies by 8 and checks you get the same value as the original parameter.

---
In reply to: [273996915](https://github.com/dotnet/wcf/pull/3538#discussion_r273996915) [](ancestors = 273996915)